### PR TITLE
retry only on real network errors

### PR DESCRIPTION
**Pull request by marco-jantke**
Now retries only happen when actual network errors occur and not only
anymore based on the HTTP status code. This is because the backend could
also send this status codes as their normal interface and in that case
we don't want to retry.

This PR should solve the issue [1104](https://github.com/containous/traefik/issues/1104)

### CODE REVIEW

File: middlewares/retry_test.go [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+
+		t.Run(tcName, func(t *testing.T) {
+			t.Parallel()
+
+			var httpHandler http.Handler
+			httpHandler = &networkFailingHTTPHandler{failAtCalls: tc.failAtCalls}
+			httpHandler = NewRetry(tc.retryCount, httpHandler)
+
+			recorder := httptest.NewRecorder()
+			req, err := http.NewRequest("GET", "http://localhost:3000/ok", ioutil.NopCloser(nil))
```
Comment (timoreimann) [inline middlewares/retry_test.go]: you could `s/"GET"/http.MethodGet/` here.

In some other parts of the PR too.

File: middlewares/retry_test.go [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+			if err != nil {
+				panic(fmt.Sprintf("NewRequest(): %s", err))
+			}
+
+			httpHandler.ServeHTTP(recorder, req)
+
+			if tc.responseStatus != recorder.Code {
+				t.Errorf("wrong status code, expected %v: got %v", tc.responseStatus, recorder.Code)
```
Comment (timoreimann) [inline middlewares/retry_test.go]: In Go: _got_ first, _expected_/_actual_ second. E.g.,

_"got status code %v, expected %v"_

In other parts of this PR too.

File: middlewares/retry_test.go [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+			}
+		})
+	}
+}
+
+// An http.Handler implementation you can use to test retries.
```
Comment (timoreimann) [inline middlewares/retry_test.go]: Should start with the type you are describing, i.e.:

_networkFailingHTTPHandler is an http.Handler implementation you can use to test retries._

File: server/errorhandler.go [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+	"net/http"
+
+	"github.com/containous/traefik/middlewares"
+)
+
+// RecordingErrorHandler is an error handler, implementing the vulcand/oxy
+// error handler interface, which is recording network errors by setting the
+// request context value with the key middlewares.DefaultNetErrCtxKey to true.
+// This enables the retry middleware to only retry on actual network errors.
+type RecordingErrorHandler struct{}
```
Comment (timoreimann) [inline server/errorhandler.go]: Since we only care about implementing `ServeHTTP(http.ResponseWriter, *http.Request, error)`, couldn't `RecordingErrorHandler` simply be defined as a function with that signature instead of an empty struct?

File: server/errorhandler.go [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+// NewRecordingErrorHandler creates and returns a new instance of RecordingErrorHandler.
+func NewRecordingErrorHandler() *RecordingErrorHandler {
```
Comment (timoreimann) [inline server/errorhandler.go]: As the implementation of `NewRecordingErrorHandler` isn't doing anything besides returning a pointer to an empty struct, couldn't we just drop it and have callers do it instead?

File: server/errorhandler.go [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+	}
+
+	w.WriteHeader(statusCode)
+	w.Write([]byte(http.StatusText(statusCode)))
+}
+
+func recordNetError(req *http.Request) {
+	val := req.Context().Value(middlewares.DefaultNetErrCtxKey)
+
+	if netErrorOccured, ok := val.(*bool); ok {
```
Comment (timoreimann) [inline server/errorhandler.go]: Does this explode if `val` is `nil` (i.e., the value does not exist), or will Go handle this fine with `ok` being `false`?

File: middlewares/retry_test.go [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+	failAtCalls []int
+	callNumber  int
+}
+
+func (handler *networkFailingHTTPHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	handler.callNumber++
+
+	for _, failAtCall := range handler.failAtCalls {
+		if handler.callNumber == failAtCall {
+			val := r.Context().Value(DefaultNetErrCtxKey)
```
Comment (timoreimann) [inline middlewares/retry_test.go]: How about exporting `server.recordNetError` and calling it here?

File: server/errorhandler_test.go [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+
+		t.Run(test.name, func(t *testing.T) {
+			t.Parallel()
+
+			recorder := httptest.NewRecorder()
+
+			netErrorOccured := false
+			req, err := http.NewRequest("GET", "http://localhost:3000/any", nil)
+			if err != nil {
+				panic(fmt.Sprintf("NewRequest(): %s", err))
```
Comment (timoreimann) [inline server/errorhandler_test.go]: A `MustNewRequest` helper function might become convenient at this point.

File: server/errorhandler.go [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+	"net"
+	"net/http"
+
+	"github.com/containous/traefik/middlewares"
+)
+
+// RecordingErrorHandler is an error handler, implementing the vulcand/oxy
+// error handler interface, which is recording network errors by setting the
+// request context value with the key middlewares.DefaultNetErrCtxKey to true.
+// This enables the retry middleware to only retry on actual network errors.
```
Comment (timoreimann) [inline server/errorhandler.go]: IMHO we should also document the type of errors it maps to 502 and 504, respectively.

File: server/errorhandler_test.go [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+			err:                io.EOF,
+			wantHTTPStatus:     http.StatusBadGateway,
+			wantNetErrRecorded: true,
+		},
+		{
+			name:               "custom error",
+			err:                errors.New("any error"),
+			wantHTTPStatus:     http.StatusInternalServerError,
+			wantNetErrRecorded: false,
+		},
```
Comment (timoreimann) [inline server/errorhandler_test.go]: Are we missing a test case for `err == nil`?

File: server/errorhandler.go (hunk 6 above)
Comment (marco-jantke) [inline server/errorhandler.go]: `go` is handling this fine. The `ctx.Value()` guarantees a return value of type `interface{}`, so we can do a type assertion without danger.

File: middlewares/retry.go [hunk 11]
```diff
@@ -56,9 +63,12 @@ func (retry *Retry) ServeHTTP(rw http.ResponseWriter, r *http.Request) {
 	}
 }
 
-func isNetworkError(status int) bool {
-	return status == http.StatusBadGateway || status == http.StatusGatewayTimeout
-}
+// NetErrorCtxKey is a custom type that is used as key for the request context.
+// Using an own type we avoid potential collisions with other keys that may have the same string name.
```
Comment (marco-jantke) [inline middlewares/retry.go]: Actually its even a best practice to not export the type is not necessary, to have real safety. Though I decided to have the error handler in the server package still, as for me it makes more sense there, which requires to export the type.

File: server/errorhandler_test.go (hunk 8 above)
Comment (marco-jantke) [inline server/errorhandler_test.go]: There is a util function in `httptest.NewRequest` that will internally panic if the creation of the `NewRequest` fails. I think this would be good to use for the tests.

File: server/errorhandler_test.go (hunk 10 above)
Comment (marco-jantke) [inline server/errorhandler_test.go]: Yes, better to also add a test for this. Doing.

File: middlewares/retry.go [hunk 12]
```diff
@@ -56,9 +63,12 @@ func (retry *Retry) ServeHTTP(rw http.ResponseWriter, r *http.Request) {
...
+type NetErrorCtxKey string
```
Comment (gottwald) [inline middlewares/retry.go]: From my learnings about context I remember that the key should be a custom type as you do here, but it should also be non exported, since it's something that should be package internal.
Obviously this is not straightforward here because the server package needs to signal something to the middleware package.
But instead of exporting low level package stuff, should the retry middlewares package maybe export some functions to deal with this and keep the context key type private?
Looks like you already have some kind of function in the server package called: `recordNetError`.
This should probably be moved to the middlewares package.
Also `recordNetError` doesn't need to get passed a request, a context would be sufficient.

Comment (timoreimann) [inline middlewares/retry.go]: An Internal package _might_ be helpful here too.

File: middlewares/retry.go [hunk 13]
```diff
@@ -42,10 +43,16 @@ func (retry *Retry) ServeHTTP(rw http.ResponseWriter, r *http.Request) {
 	}
 	attempts := 1
 	for {
+		netErrorOccured := false
```
Comment (marco-jantke) [inline middlewares/retry.go]: Thanks, fixed.

File: middlewares/retry.go [hunk 14]
```diff
@@ -42,10 +43,16 @@ func (retry *Retry) ServeHTTP(rw http.ResponseWriter, r *http.Request) {
...
+		// We pass in a pointer to netErrorOccured so that we can set it to true on network errors
+		// that happen when proxying the http requests to the backends.
```
Comment (marco-jantke) [inline middlewares/retry.go]: Ok, I extended the comment.

File: middlewares/retry.go [hunk 15]
```diff
@@ -42,10 +43,16 @@ func (retry *Retry) ServeHTTP(rw http.ResponseWriter, r *http.Request) {
...
+		newCtx := context.WithValue(r.Context(), DefaultNetErrCtxKey, &netErrorOccured)
+
 		recorder := NewRecorder()
 		recorder.responseWriter = rw
-		retry.next.ServeHTTP(recorder, r)
-		if !isNetworkError(recorder.Code) || attempts >= retry.attempts {
+
+		retry.next.ServeHTTP(recorder, r.WithContext(newCtx))
+		if netErrorOccured == false || attempts >= retry.attempts {
```
Comment (marco-jantke) [inline middlewares/retry.go]: Ok, done.

File: middlewares/retry.go (hunk 11 above)
Comment (marco-jantke) [inline middlewares/retry.go]: I refactored the implementation now, so that the knowledge about how to interact with the context completely to the retry middleware and therefore even can use a private type.

File: middlewares/retry_test.go (hunk 1 above)
Comment (marco-jantke) [inline middlewares/retry_test.go]: Thanks, fixed at all places.

File: middlewares/retry_test.go (hunk 2 above)
Comment (marco-jantke) [inline middlewares/retry_test.go]: Yes.. Fixed it. It seems to be programmed into my fingertips.

File: middlewares/retry_test.go (hunk 3 above)
Comment (marco-jantke) [inline middlewares/retry_test.go]: Ok, done.

File: middlewares/retry_test.go (hunk 7 above)
Comment (marco-jantke) [inline middlewares/retry_test.go]: As said before, I refactored this implementation, so the implementation for recording the network error is now in the middleware package and the tests make more sense.

File: server/errorhandler.go (hunk 9 above)
Comment (marco-jantke) [inline server/errorhandler.go]: I added a general comment that it will map to the proper status codes. I fear that being to explicit in the comment will bring up the danger, that the implementation and the comment will drift apart at some time. And when people read the comment, they probably also can read the code.

File: server/errorhandler.go (hunk 4 above)
Comment (marco-jantke) [inline server/errorhandler.go]: After my refactoring it has now a property for the NetErrorRecorder, so its not possible as function anymore.

File: server/errorhandler.go (hunk 5 above)
Comment (marco-jantke) [inline server/errorhandler.go]: After my refactoring it has a property now and I would leave the `New..` function.

File: middlewares/retry.go (hunk 12 above)
Comment (marco-jantke) [inline middlewares/retry.go]: Thanks for the hint, I refactored it now in this way. Now the logic for mutating the context is solely in the retry middleware and therefore we can use a private type 🎉

**Comment by marco-jantke** [comment]
@timoreimann and @gottwald thanks for the review, I worked your feedback into the PR.

File: middlewares/retry.go (hunk 12 above)
Comment (marco-jantke) [inline middlewares/retry.go]: @timoreimann I am not sure how this should look like to be honest. Also I am hesitant to change the middleware structure in this PR, as imho it would be separate concern. Anyway I think for a long term it could be beneficial to bring more structure into the middleware package.

File: middlewares/retry_test.go [hunk 16]
```diff
@@ -0,0 +1,92 @@
...
+
+			if tc.responseStatus != recorder.Code {
+				t.Errorf("got status code %v, wanted %v", recorder.Code, tc.responseStatus)
+			}
+		})
+	}
+}
+
+func TestRecordNetError(t *testing.T) {
+	boolNetErrorOccurred := false
```
Comment (marco-jantke) [inline middlewares/retry_test.go]: A small hint for reviewers, I tried to build this as a table driven test at first, but I was not able to get it working in a straightforward way, as it would require me to specify _one_ type for the initial context value I am passing in. This would have to be an `interface{}` in my case, but then the type assertion on a `*bool` in the `DefaultRecordNetError` does not work anymore.

File: middlewares/retry.go [hunk 17]
```diff
@@ -56,8 +63,34 @@ func (retry *Retry) ServeHTTP(rw http.ResponseWriter, r *http.Request) {
...
-	return status == http.StatusBadGateway || status == http.StatusGatewayTimeout
+// netErrorCtxKey is a custom type that is used as key for the context.
+type netErrorCtxKey string
+
+// defaultNetErrCtxKey is the actual key which value is used to record network errors.
+var defaultNetErrCtxKey netErrorCtxKey = "NetErrCtxKey"
+
+// NetErrorRecorder is an interface to record net errors.
+type NetErrorRecorder interface {
+	// RecordNetError can be used to signal the retry middleware, that an network error happened
```
Comment (timoreimann) [inline middlewares/retry.go]: No comma before `that`.

File: middlewares/retry_test.go (hunk 16 above)
Comment (timoreimann) [inline middlewares/retry_test.go]: Then let's have three separate test functions so that we can easily see which test failed, can easily rerun them, etc.

File: middlewares/retry.go [hunk 18]
```diff
@@ -42,10 +43,16 @@ func (retry *Retry) ServeHTTP(rw http.ResponseWriter, r *http.Request) {
 	}
 	attempts := 1
 	for {
+		netErrorOccurred := false
+		// We pass in a pointer to netErrorOccurred so that we can set it to true on network errors
+		// when proxying the http requests to the backends. This happens in the custom RecordingErrorHandler.
```
Comment (timoreimann) [inline middlewares/retry.go]: s/http/HTTP/

File: server/errorhandler.go [hunk 19]
```diff
@@ -0,0 +1,40 @@
...
+	"io"
+	"net"
+	"net/http"
+
+	"github.com/containous/traefik/middlewares"
+)
+
+// RecordingErrorHandler is an error handler, implementing the vulcand/oxy
+// error handler interface, which is recording network errors by using the netErrorRecorder.
+// In addition it sets a proper http status code and body, depending on the type of error occurred.
```
Comment (timoreimann) [inline server/errorhandler.go]: s/http/HTTP/

File: middlewares/retry.go (hunk 18 above)
Comment (marco-jantke) [inline middlewares/retry.go]: Thanks, done.

File: middlewares/retry.go (hunk 17 above)
Comment (marco-jantke) [inline middlewares/retry.go]: Thanks, done.

File: server/errorhandler.go (hunk 19 above)
Comment (marco-jantke) [inline server/errorhandler.go]: Thanks, done.

File: middlewares/retry_test.go (hunk 16 above)
Comment (marco-jantke) [inline middlewares/retry_test.go]: Ok, done.

**Comment by marco-jantke** [comment]
Ok, worked in feedback again.

File: server/errorhandler.go [hunk 20]
```diff
@@ -0,0 +1,40 @@
...
+// NewRecordingErrorHandler creates and returns a new instance of RecordingErrorHandler.
+func NewRecordingErrorHandler(recorder middlewares.NetErrorRecorder) *RecordingErrorHandler {
+	return &RecordingErrorHandler{recorder}
+}
+
+func (eh *RecordingErrorHandler) ServeHTTP(w http.ResponseWriter, req *http.Request, err error) {
+	statusCode := http.StatusInternalServerError
+
+	if e, ok := err.(net.Error); ok {
+		eh.netErrorRecorder.RecordNetError(req.Context())
```
Comment (gottwald) [inline server/errorhandler.go]: nit-pick: This stutters a bit, maybe call the method `Record` instead?

Comment (marco-jantke) [inline server/errorhandler.go]: Thanks for pointing this out, changed it.

**Comment by marco-jantke** [comment]
@emilevauge or @vdemeester can you have a look?

**Comment by ldez** [comment]
Due to SemaphoreCI, I close and reopen the PR.

File: middlewares/retry.go [hunk 21]
```diff
@@ -56,8 +63,34 @@ func (retry *Retry) ServeHTTP(rw http.ResponseWriter, r *http.Request) {
...
+
+// NetErrorRecorder is an interface to record net errors.
+type NetErrorRecorder interface {
+	// Record can be used to signal the retry middleware that an network error happened
+	// and therefore the request should be retried.
+	Record(ctx context.Context)
+}
+
+// NetErrorRecorderFunc is an util to be able to implement the NetErrorRecorder without the need of a struct.
+type NetErrorRecorderFunc func(ctx context.Context)
```
Comment (emilevauge) [inline middlewares/retry.go]: What is the need of doing this ? Why don't you want to use a `struct` here ?

File: middlewares/retry.go [hunk 22]
```diff
@@ -56,8 +63,34 @@ func (retry *Retry) ServeHTTP(rw http.ResponseWriter, r *http.Request) {
...
+// Record is there to satisfy the NetErrorRecorder interface on a NetErrorRecorderFunc.
+func (f NetErrorRecorderFunc) Record(ctx context.Context) {
+	f(ctx)
+}
+
+// DefaultRecordNetError is recording network errors by setting the context value for the defaultNetErrCtxKey to true.
+func DefaultRecordNetError(ctx context.Context) {
+	val := ctx.Value(defaultNetErrCtxKey)
+
+	if netErrorOccurred, ok := val.(*bool); ok {
```
Comment (emilevauge) [inline middlewares/retry.go]: Not easy to read.

File: server/errorhandler.go [hunk 23]
```diff
@@ -0,0 +1,40 @@
+package server
+
+import (
+	"io"
+	"net"
+	"net/http"
+
+	"github.com/containous/traefik/middlewares"
+)
+
```
Comment (emilevauge) [inline server/errorhandler.go]: You will probably get issues with websockets here.
You should add 
```
var (
	_ Stateful = &RecordingErrorHandler{}
)
```

File: middlewares/retry.go (hunk 21 above)
Comment (marco-jantke) [inline middlewares/retry.go]: The implementation has no need of any field. I originally had a `struct` here without fields, but got a hint in a previous review here that a function would suffice. Therefore I added this type in order to able to implement the `NetErrorRecorder` interface on a function.

Comment (marco-jantke) [inline middlewares/retry.go]: This said I am fine with both implementations. An empty `struct` is a bit less code, but in some way unnecessary. As you prefer :)

File: server/errorhandler.go (hunk 23 above)
Comment (marco-jantke) [inline server/errorhandler.go]: Hmm it has none of the capabilities the `Stateful` interface expects. Maybe the naming is rather not perfect then? It is not an extension of `ResponseWriter` or whatsoever, its an implementation of the oxy `ErrorHandler` interface. Its called `Recording` because it calls the `NetErrorRecorder` which in our case will manipulate the request context. So I think this is unnecessary here. Any suggestion for naming improvements?

File: middlewares/retry.go (hunk 22 above)
Comment (marco-jantke) [inline middlewares/retry.go]: I renamed `ok` now to `isBoolPointer`. Not sure how to improve readability in another way.

**Comment by marco-jantke** [comment]
@emilevauge please have a look again, incorporated/answered your questions.

File: middlewares/retry.go (hunk 21 above)
Comment (emilevauge) [inline middlewares/retry.go]: If that's OK for you, I would prefer an empty struct, which is easier to read, and with the same result :)

File: server/errorhandler.go [hunk 24]
```diff
@@ -0,0 +1,40 @@
...
+	statusCode := http.StatusInternalServerError
+
+	if e, ok := err.(net.Error); ok {
+		eh.netErrorRecorder.Record(req.Context())
+		if e.Timeout() {
+			statusCode = http.StatusGatewayTimeout
+		} else {
+			statusCode = http.StatusBadGateway
+		}
+	} else if err == io.EOF {
```
Comment (emilevauge) [inline server/errorhandler.go]: Reading this part again, not sure this is the right way to go. We could get an EOF error *not* due to network right ?

File: server/errorhandler.go (hunk 23 above)
Comment (emilevauge) [inline server/errorhandler.go]: You're right, successfully tested your code with websocket :+1:

File: middlewares/retry.go (hunk 21 above)
Comment (marco-jantke) [inline middlewares/retry.go]: No problem, I changed it to a struct now.

File: server/errorhandler.go (hunk 24 above)
Comment (marco-jantke) [inline server/errorhandler.go]: About this one I am not 100% sure whether its right to retry here. I am currently trying to investigate use-cases when an io.EOF can happen. For sure it can be due to a network connection problem, but also when the server closes the connection by intend. Maybe the server would do so because of too much load and we should rather try another one? Anyway I think there is no way to vary in between those cases and thats why in general I would rather also retry those request.

Comment (marco-jantke) [inline server/errorhandler.go]: @emilevauge WDYT?

Comment (emilevauge) [inline server/errorhandler.go]: @timoreimann any thoughts on this ?

Comment (timoreimann) [inline server/errorhandler.go]: @emilevauge @marco-jantke the net package does [distinguish between an an `ErrUnexpectedEOF` and `EOF` state](https://golang.org/pkg/io/#pkg-variables). Unfortunately, it's [not used by net/http](https://github.com/golang/go/issues/18055).

Even if we considered this to be changeworthy, IMHO we should not do so along this PR which maintains the current oxy error handler behavior except for the response code logic. We can have a separate issue to drive a potential improvement in this focused regard.

Comment (emilevauge) [inline server/errorhandler.go]: @marco-jantke I agree with @timoreimann on this, we should deal with this issue in another PR. The fact that it is not used in the default go http client leads to discussions :)

Comment (marco-jantke) [inline server/errorhandler.go]: I agree, so we handled all the questions/feedback in this PR I think :)

**Comment by marco-jantke** [comment]
@timoreimann @emilevauge you are ok with me doing a rebase already to resolve the conflicts?

**Comment by timoreimann** [comment]
@marco-jantke I think it's okay to do that now.

**Comment by marco-jantke** [comment]
Ok, everything should be ready now.

**Comment by marco-jantke** [comment]
@emilevauge ping :)

**Comment by ldez** [comment]
@marco-jantke We browse all PRs regularly, you don't need to ping someone.
